Point on feature final

Newmanjack opened the pull request:
These changes aim to implement and test a new feature called pointOnFeature, which returns a GeoJSON Feature<Point> guaranteed to be on the surface of the input feature. This implementation replicates Turf.js functionality and supports multiple geometry types.

Newmanjack committed:
Add pointOnFeature functionality to find points on GeoJSON features

Newmanjack committed:
Update pointOnFeature tests and export functionality in turf.dart

Newmanjack committed:
Move point_on_feature test to components directory and improve test organization

Newmanjack committed:
Add point_on_feature benchmark for performance testing

lukas-h commented on benchmark/point_on_feature_benchmark.dart:
```diff
@@ -0,0 +1,65 @@
+import 'package:turf/turf.dart';
+
+// Create some test features for benchmarking
+var point = Feature(
```
use final instead of var if value is only assigned once

lukas-h commented on lib/src/point_on_feature.dart:
```diff
@@ -0,0 +1,214 @@
+import 'dart:math' as math;
+import 'package:geotypes/geotypes.dart'; // We still need the GeoJSON types, as they're used throughout the package
+
+/// Returns a Feature<Point> that represents a point guaranteed to be on the feature.
```
make sure we are using the proper dart doc comment syntax (wrapping types in [ ] e.g.)

https://dart.dev/effective-dart/documentation

lukas-h commented on lib/src/point_on_feature.dart:
```diff
@@ -0,0 +1,214 @@
+  }
+  
+  // Unsupported input type.
+  return null;
```
let's make this function non-nullable and throw an exception instead. what do you think?

aardrop replied:
Agreed, error handling with messaging including "unsupported input type (type of input)"

lukas-h commented on lib/src/point_on_feature.dart:
```diff
@@ -0,0 +1,214 @@
+}
+
+/// Calculates the arithmetic centroid of a Polygon's outer ring.
+Point calculateCentroid(Polygon polygon) {
```
we already have a centroid function – can you check if we can use that instead? This would allow us to have more code sharing

lukas-h replied:
less code duplication

aardrop replied:
So turf has updated a lot recently, it would be good to compare reference code for the optimal version. Turf JS also breaks out many methods to be isolated so I think we need to decide if we want to do that or anticipate pulling the whole package and trusting the compiled to remove unused code.

lukas-h commented on lib/src/point_on_feature.dart:
```diff
@@ -0,0 +1,214 @@
+}
+
+/// Calculates a representative midpoint on a LineString.
+Feature<Point> _midpointOnLine(LineString line, Map<String, dynamic>? properties) {
```
there should be a similar function for midpoint that we could use

lukas-h commented on lib/src/point_on_feature.dart:
```diff
@@ -0,0 +1,214 @@
+}
+
+/// Checks if a point is inside a polygon using a ray-casting algorithm.
+bool _pointInPolygon(Position point, Polygon polygon) {
```
there is also already an extra pip library that we might be able to use

lukas-h reviewed (changes requested):
I found a lot of potential to share code with already existing functions of this lib. Can you check that out? :)

Newmanjack committed:
Added toWGS84 and toMercator as member functions of the coordinate types

Newmanjack committed:
Improve documentation for coordinate projection functions to follow Dart standards

Newmanjack committed:
Add point_on_feature examples with visualizations

Newmanjack committed:
Refactor pointOnFeature to use modular methods and improve error handling.

- Replace custom implementations with existing specialized modules
- Use turf_pip package directly for point-in-polygon checks
- Use turf_centroid, turf_midpoint, turf_area and turf_length libraries
- Make function non-nullable with proper error handling
- Update var to final in benchmark file and across codebase
- Update tests to work with geodesic midpoint calculations
- Improve documentation and code quality

Newmanjack committed:
Remove coordinate system conversion changes from helpers.dart

This reverts changes made in commits:
- 41bfee542794267a3fb656158a6a783e26753e44
- a343ee78fbc912e3a5d019911fc8735b39e48ee7

These coordinate conversion features will be handled in a separate PR (oWSG84_and_toMercator_unit)

Newmanjack committed:
Update helpers_test.dart with version from main

GuannanMa committed:
use final in helper utilities

lukas-h reviewed (approved):
Approved!

lukas-h commented:
@GuannanMa please just mention this implementation in the Progress.md, then you can squash-merge yourself without me in the loop!

CC @hamishdgx

GuannanMa committed:
clean analyzer issues in point on feature work

GuannanMa committed:
Merge branch 'main' into point_on_feature_final

GuannanMa committed:
fix analyzer issues after syncing main
